### DISCUSSION TITLE
[solved] what is [FIN, ACK] packet.

### DISCUSSION THREAD

**Discussion by rubidam** [ANLAB-KAIST/KENSv3#116] [Project Q&A 2. Transport Layer (TCP)] [answered]
When I test backlog1 case with 'wireshark', I saw that server receive [fin, ack] packet from client. by searching, [fin, ack] packet is transmitted when client or server call 'shutdown' function for the 'half-close'. below is my question.

1. Is this a situation that client call 'shutdown' function?
**-> I saw 'rfc793 document file', maybe it means that client call 'close' function. so ignore 2,3 question.**
2. The 'shutdown' function is not executed in the testhandshake.cpp file, but I am not sure whether the [fin, ack] packet came because we did something wrong during the 3-way handshaking process, or whether this packet came normally by calling shutdown somewhere.
3. If the answer to number 1 is yes, can the server just send an ack packet?

4. according to the results of wireshark and the results that I print, the client close immediately after completing 3-way handshaking with the server. The server is not yet accepting. At this time, I don't understand why the server should accept this client.

I'll attach the wireshark results together.

-------------------------------------------------------------------------------------------------------------------------------------------------
안녕하세요. backlog1 테스트 케이스를 wireshark를 이용하여 패킷 분석중에 클라이언트가 서버에게 [fin, ack] 패킷을 보내는 것을 관찰했는데 이 부분이 정확히 이해가 되지 않아 질문 올립니다. [fin, ack] 패킷은 half-close를 위해 shutdown함수를 부르면 전송되는 패킷이라고 구글링을 통해 알게 되었습니다.

1. [fin, ack] 패킷이 클라이언트가 shutdown 함수를 부른 상황인가요?
**-> rfc793 document 파일을 보았는데 client가 close를 부른 상황인 것 같습니다. 2,3번 질문은 무시하면 될 것 같습니다.**
2. testhandshake.cpp 파일에서 shutdown이라는 함수를 실행하지 않는데 저희가 3-way handshaking 과정에서 잘못된 행동을 해서 [fin, ack] 패킷이 온건지, 어딘가에서 shutdown을 불러서 이 패킷이 정상적으로 온 상황인지 잘 모르겠습니다.
3. 1번의 대답이 yes 라면, 서버의 행동은 그냥 ack패킷만 보내면 되나요?
4. backlog1 케이스의 wireshark의 결과와 제가 printf한 결과를 보면 client는 server와 3-way handshaking을 완료한 후 즉시 close을 하게 됩니다. 아직 server는 accept를 하지 않은상태이고요. 이때 server는 왜 이 클라이언트에 대해 accept를 해야하는지 잘 이해가 되지 않습니다.

wireshark 결과를 함께 첨부하겠습니다.

![image](https://user-images.githubusercontent.com/44969740/194738847-ab2c905e-6cf0-41b1-8a02-dbdfd00fdc86.png)

**Comment by rubidam** [comment, marked as answer]
I solved this problem

**Reply by rubidam** [reply]
1. [Fin,ack] packet is sent by calling close() function at client.
위 패킷은 클라이언트가 close함수를 불러서 전송된 상황이 맞는거 같습니다.
2. My expected answer of question 4 is following
  - it is half close state so welcome socket can establish connection
  - when server make connection socket, connection socket should know that client socket already call close(). 
일단 printf로 확인한 함수실행 순서랑 wireshark로 예상한 순서랑은 안맞을수도 있어서 accept전에 close를 불렀다는 제 예상이 틀렸을수도 있습니다. 제 예상이 맞다는 전제로 푼다면 클라이언트가 close를 불렀을때는 accept전이라서 welcome socket이 finack패킷을 받는거 같습니다. 그럼 welcome socket은 ack로 응답하죠. 근데 이때는 half close상태라서 accept를 할 수 있고, connection socket이 이후에 만들어집니다. 그리고 server의 connection socket이 close를 부르면 finack packet과 ack packet을 주고받으면서 완전히 커넥션이 끝난다고 생각했습니다. 
영어로 설명드리기가 너무 어려워서 한글로 설명드렸습니다... 제설명이 도움이 됐으면 좋겠네요. 혹시 이해가 안되시거나 도움 필요하시면 답글이나 메일 주셔도됩니다.

**Comment by whkim0** [comment]
Let me skip Q2 and 3 then.
1\. Yes, client is initiating the close.
4\. I couldn't fully understand your question:
> ... the client close immediately after completing 3-way handshaking with the server. The server is not yet accepting.
 At this time, I don't understand why the server should accept this client.

I guess you want to say the server "establishes a connection with the client" by "accepting the client". 
The connection is established between 10.0.1.4 and 192.168.0.7 when the three-way handshake is done.

Can you elaborate more about "The server is not yet accepting" if you want, so that I can give you a better answer?
